### ISSUE TITLE
Cache table population

### ISSUE THREAD

**Issue by johnvanbreda** [Indicia-Team/warehouse#18]
Currently, of the 4 cache tables that are populated for reporting simplicity and performance (cache_occurrences, cache_taxa_taxon_lists, cache_taxon_searchterms and cache_termlists_terms), only cache_occurrences is populated on the fly as records are added or updated. 

This proposal is to also populate the other cache tables on the fly as data are added, updated and deleted. 

There are a number of advantages:
* No delay between a species or term being added to the warehouse and that species or term being available for use in the data entry forms. Currently the scheduled tasks need to happen on the warehouse first.
* As the cache tables will always contain the latest version of the data, there would be no excuse for directly querying the underlying tables (e.g. joining taxa_taxon_lists, taxa etc to get various names for a species). This means that we can review all the built in views to make sure that they are using the cache_* tables rather than underlying tables and we can also do the same for all reports. This should improve performance.
* We could streamline the tasks that run when the scheduled_tasks run as the cache_* tables will no longer require any updating.
* As there would be less dependencies in the database on the term table, script version_0_9_0\201303041000_term_field_size.sql could be removed and a new script added to the end of the script sequence which performs this field size change by direct ALTER sql rather than a dodgy update to the pg_attribute table. Although not a big issue, this particular script does make the installation process somewhat more confusing for new warehouses.

The disadvantages:
* The process of uploading new species data or termlist data would be slightly slower. However this is an admin task and not something that happens that frequently.

**Comment by JimBacon** [comment]
As you say, adding terms and taxa are an admin task which are performed infrequently. Updating the cache tables immediately on those occasions is probably going to occupy far fewer processor cycles overall than checking for changes every time the scheduled tasks runs. One possible reservation might be the performance impact if performing a big update from the UK Species Inventory.

I see no reason, however, to make the use of cache tables in reports contingent on this change as the delay introduced by scheduled tasks is hardly significant. Ensuring all reports use the cache tables is the principal way in which we can have an impact on performance.

The following article makes interesting reading as an alternative way to maintain a cache.
http://tech.jonathangardner.net/wiki/PostgreSQL/Materialized_Views 
I can''t think that it offers any advantage over what we are doing at present.
However, one way to improve performance might be to host the database on a different machine from the web server. If the database is then responsible for maintaining the cache rather than a script run by the web server, the responsiveness of the web server might improve for write operations.

**Comment by JimBacon** [comment]
Also new to me is http://www.postgresql.org/docs/9.3/static/rules-materializedviews.html but this only seems to generate a snapshot which is not what we are interested in.

**Comment by johnvanbreda** [comment]
See #36 for separate issue to cover updating the report files - as you say there is no hard-dependency on the modification proposed here.

**Comment by johnvanbreda** [comment]
Reviewing the options available:

1. Materliazed view (probably eager?). Problem with this approach is that they are based on a single view query, but the script required to build cache_occurrences for example is very complex. Therefore it would be difficult to rewrite in a single query and if it were possible I would expect such a complex query to be less efficient than a script broken down into several steps.
2. Use database triggers to maintain the cache_* tables. This would be:
   * Good in that the logic is encapsulated in the database layer (performance and robustness)
   * As the trigger would fire on every single update of the underlying tables, it might be that certain optimisations are not possible. For example saving a single sample containing 10 occurrences would fire 11 triggers - whereas in a PHP script you can be clever and perform a single update. This could be significant during imports of data.
3. PHP server side script fired by My_ORM updates - this currently happens for cache_occurrences but not the other cache tables. The original suggestion was to switch to this approach for all tables. It is perhaps less robust than a trigger but potentially more optimal for any updates of multiple records, e.g. record lists or imports.
4. PHP server side script fired by scheduled_tasks. This allows optimisations in that a whole chunk of records can be done at once. However you are repeatedly scanning for updates which is a performance hit even when no records are added. Also there is a small issue of the time delay between database changes and the cache tables updating. 

Personally I think the best approach is still unclear to me, as we need more information about the relative performance impacts of the different approaches. So I'd be tempted to proceed with #36 for now and revisit this with a proper exploration of the options later.
